pkg/pillar/hypervisor/xen.go

```diff
@@ -583,6 +580,44 @@ func (ctx xenContext) PCIRelease(long string) error {
 	return nil
 }
 
+//IsDomainHealthy: Verifies if a domain is up and running by its state. Domain can have the following states:
+//r - currently running
+//b - blocked, and not running or runnable
+//p - paused
+//s - a shutdown command has been sent, but the domain isn't dying yet
+//c - the domain has crashed
+//d - the domain is dying, but hasn't properly shut down or crashed
+func (ctx xenContext) IsDomainHealthy(domid int) bool {
+	domainState := ""
+	cmd := "xl"
+	args := []string{
+		"list",
+		strconv.Itoa(domid),
+	}
+	stdoutStderr, err := wrap.Command(cmd, args...).CombinedOutput()
+	if err != nil {
+		//domain is not present
+		log.Errorln("IsDomainHealthy: xl list failed ", err)
+		log.Errorln("IsDomainHealthy: xl list output ", string(stdoutStderr))
+		return false
+	}
+	//Removing all extra space between column result and split the result as array.
+	xlDomainResult := regexp.MustCompile(`\s+`).ReplaceAllString(strings.Split(string(stdoutStderr), "\n")[1], " ")
+	//Domain's status is 5th column in xl list <domain> result
+	domainState = strings.Split(xlDomainResult, " ")[4]
+	//Removing all unset state bits represented by "-"
+	domainState = strings.ReplaceAll(domainState, "-", "")
+	log.Infof("IsDomainHealthy: domain: %d domainState: %s.", domid, domainState)
+	//In case of more that 1 (logically possible) domain state, will consider the last state.
+	lastState := domainState[len(domainState)-1:]
```

Comment:
len(domainState) can be zero here when it is "-----" above, so you will panic. You had that check earlier.
return false in that case.

Reply:
Added.